# Pytorch - Redes neuronales


Se realizara la misma prediccion de one vs all pero esta vez usando pytorch para ver la diferencia
como se vio anteriormente el dataset usado es personalidad.csv que no tiene una fila de enteros pero eso se cambia con funciones de pandas, los valores que tomaran de la columna personality son:

Introvert = 0

Extrovert = 1

Ambivert = 2

In [ ]:
import torch
import pandas as pd
from sklearn.datasets import fetch_openml
import numpy as np

Utilizando la clase Sequentail. Esta clase nos permite definir una secuencia de capas, que se aplicarán de manera secuencial (las salidas de una capa serán la entrada de la siguiente)

- Datos de entrada 29, ya que se cuenta con esa cantidad de de columnas en X
- H numero de neuronas en la capa interna, este numero puede variar segun la demanda
- Datos de salida, son 3 porque en este modelo de one vs all se tienen en y 3 valores distintos que ya vimos

In [44]:
D_in, H, D_out = 29, 100, 3

model = torch.nn.Sequential(
    torch.nn.Linear(D_in, H),
    torch.nn.ReLU(),
    torch.nn.Linear(H, D_out),
)

Hacemos la prueba para saber si funciona la red neuronal
el resultado devuelto se interpreta como, 600 ejemplos con 3 valores cada 1

In [45]:
outputs = model(torch.randn(600, 29))
outputs.shape

torch.Size([600, 3])

In [46]:
print(outputs[0][:])

tensor([-0.1066,  0.6211,  0.1195], grad_fn=<SliceBackward0>)


In [47]:
model

Sequential(
  (0): Linear(in_features=29, out_features=100, bias=True)
  (1): ReLU()
  (2): Linear(in_features=100, out_features=3, bias=True)
)

Despues de la verificamos cargamos el modelo

In [48]:
data = pd.read_csv('personalidad.csv', delimiter=',')

# Mapear y a valores enteros
mapping = {'Introvert': 0, 'Extrovert': 1, 'Ambivert': 2}
data["personalidad"] = data["personality_type"].map(mapping)
X = data.drop(columns=["personality_type", "personalidad"])  # todas menos las columnas personality_type y personalidad
Y = data["personalidad"]
data

,personality_type,social_energy,alone_time_preference,talkativeness,deep_reflection,group_comfort,party_liking,listening_skill,empathy,creativity,...,adventurousness,reading_habit,sports_interest,online_social_usage,travel_desire,gadget_usage,work_style_collaborative,decision_speed,stress_handling,personalidad
0,Extrovert,6.794295,3.854670,8.725446,2.515151,7.097368,8.588762,6.774799,6.430132,6.142968,...,8.257134,5.270555,10.000000,9.154296,4.816422,9.191711,8.313590,8.032376,7.176905,1
1,Ambivert,6.378988,5.731157,7.029529,7.274493,4.111199,3.258248,5.550909,3.958179,6.149457,...,6.289347,5.753165,5.334303,4.683781,4.725666,5.956141,5.890619,3.158988,3.423577,2
2,Ambivert,7.459421,6.322263,3.922269,4.622261,5.343276,7.452152,9.483990,6.127654,7.032017,...,9.238784,5.250405,3.153540,5.000338,6.139166,6.033048,5.807500,4.571003,5.647480,2
3,Extrovert,6.159626,3.097837,6.019093,1.965440,7.837140,10.000000,9.436733,8.949684,8.923875,...,8.489791,5.312617,8.379936,7.601946,6.370056,5.410145,6.671781,6.600233,5.870088,1
4,Introvert,5.568462,6.986722,3.913240,9.926161,1.650483,0.362298,7.470387,6.756837,9.507803,...,3.167217,7.060235,2.333388,7.771569,5.534336,5.704598,5.832968,5.813099,3.758084,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19995,Ambivert,3.783857,4.345803,8.220063,6.297798,5.162622,5.702266,5.518356,3.420139,5.811565,...,4.355973,6.454815,2.557365,6.949292,4.176390,5.649897,5.665498,5.252077,4.297520,2
19996,Introvert,2.621406,8.152097,2.954341,8.791467,2.670252,2.931244,5.855202,5.044124,6.244061,...,2.631298,10.000000,3.697801,3.217559,3.638815,5.177155,5.106628,0.982214,5.873321,0
19997,Introvert,0.039689,8.277160,5.728101,7.790080,3.427897,0.708457,7.099321,5.784592,5.400400,...,4.071008,9.974147,2.954407,5.427296,6.270830,3.793430,5.009167,3.810962,5.678412,0
19998,Extrovert,7.756725,3.956191,8.906972,3.793775,7.608467,6.292705,6.824980,5.653421,7.347820,...,8.148780,3.840900,6.093418,7.814507,8.428745,8.736932,8.060841,8.161087,5.941793,1


Verificamos los valores de X Y

In [49]:
print("Características de entrada (X):")
print(X)
print("Etiquetas de salida (y):")
print(Y)

Características de entrada (X):
       social_energy  alone_time_preference  talkativeness  deep_reflection  \
0           6.794295               3.854670       8.725446         2.515151   
1           6.378988               5.731157       7.029529         7.274493   
2           7.459421               6.322263       3.922269         4.622261   
3           6.159626               3.097837       6.019093         1.965440   
4           5.568462               6.986722       3.913240         9.926161   
...              ...                    ...            ...              ...   
19995       3.783857               4.345803       8.220063         6.297798   
19996       2.621406               8.152097       2.954341         8.791467   
19997       0.039689               8.277160       5.728101         7.790080   
19998       7.756725               3.956191       8.906972         3.793775   
19999       3.225794               8.058594       0.000000        10.000000   

       group_comfor

In [50]:
x_2=np.array(X)
y_2=np.array(Y)

# Normalización y split
X_train = x_2[:16000]  # primeros 16k para entrenamiento
X_test  = x_2[16000:]  # últimos 4k para prueba

# Normalización 0-1
X_train = (X_train - 1) / 9
X_test  = (X_test - 1) / 9

# Convertir etiquetas a enteros
y_train = y_2[:16000].astype(np.int32)
y_test  = y_2[16000:].astype(np.int32)

Mide qué tan bien el modelo predijo la clase verdadera.

Si la probabilidad de la clase correcta es alta → pérdida baja.

Si es baja → pérdida alta.

In [51]:
# función de pérdida y derivada

def softmax(x):
    return torch.exp(x) / torch.exp(x).sum(axis=-1,keepdims=True)

def cross_entropy(output, target):
    logits = output[torch.arange(len(output)), target]
    loss = - logits + torch.log(torch.sum(torch.exp(output), axis=-1))
    loss = loss.mean()
    return loss

In [52]:
torch.cuda.is_available()

True

In [53]:
print(X)

       social_energy  alone_time_preference  talkativeness  deep_reflection  \
0           6.794295               3.854670       8.725446         2.515151   
1           6.378988               5.731157       7.029529         7.274493   
2           7.459421               6.322263       3.922269         4.622261   
3           6.159626               3.097837       6.019093         1.965440   
4           5.568462               6.986722       3.913240         9.926161   
...              ...                    ...            ...              ...   
19995       3.783857               4.345803       8.220063         6.297798   
19996       2.621406               8.152097       2.954341         8.791467   
19997       0.039689               8.277160       5.728101         7.790080   
19998       7.756725               3.956191       8.906972         3.793775   
19999       3.225794               8.058594       0.000000        10.000000   

       group_comfort  party_liking  listening_skill

Convierte tus datos de entrada (X_train, que es un numpy.ndarray) en un tensor de PyTorch y mueve ese tensor a la GPU, para que los cálculos se hagan más rápido que en CPU.

In [54]:
X_t = torch.tensor(X_train, dtype=torch.float32).cuda()
Y_t = torch.tensor(y_train, dtype=torch.long).cuda()
model = model.cuda()

# bucle entrenamiento
epochs         umero de iteraciones

lr             pasos de aprendizaje

log_each       cada cuántas iteraciones mostramos el log

l = []         lista donde guardamos los valores de la función de pérdida

In [ ]:
epochs = 350 
lr = 0.8  
log_each = 10 
l = []
for e in range(1, epochs + 1):

    # forward
    y_pred = model(X_t)

    # loss
    loss = cross_entropy(y_pred, Y_t)
    l.append(loss.item())

    # ponemos a cero los gradientes
    model.zero_grad()

    # Backprop (calculamos todos los gradientes automáticamente)
    loss.backward()

    # update de los pesos
    with torch.no_grad():
        for param in model.parameters():
            param -= lr * param.grad

    if not e % log_each:
        print(f"Epoch {e}/{epochs} Loss {np.mean(l):.5f}")

Epoch 10/350 Loss 1.00455
Epoch 20/350 Loss 0.89870
Epoch 30/350 Loss 0.77758
Epoch 40/350 Loss 0.76223
Epoch 50/350 Loss 0.72317
Epoch 60/350 Loss 0.62394
Epoch 70/350 Loss 0.54432
Epoch 80/350 Loss 0.48214
Epoch 90/350 Loss 0.43272
Epoch 100/350 Loss 0.39261
Epoch 110/350 Loss 0.35945
Epoch 120/350 Loss 0.33159
Epoch 130/350 Loss 0.30786
Epoch 140/350 Loss 0.28740
Epoch 150/350 Loss 0.26959
Epoch 160/350 Loss 0.25393
Epoch 170/350 Loss 0.24006
Epoch 180/350 Loss 0.22769
Epoch 190/350 Loss 0.21659
Epoch 200/350 Loss 0.20656
Epoch 210/350 Loss 0.19747
Epoch 220/350 Loss 0.18919
Epoch 230/350 Loss 0.18160
Epoch 240/350 Loss 0.17463
Epoch 250/350 Loss 0.16821
Epoch 260/350 Loss 0.16227
Epoch 270/350 Loss 0.15676
Epoch 280/350 Loss 0.15163
Epoch 290/350 Loss 0.14685
Epoch 300/350 Loss 0.14238
Epoch 310/350 Loss 0.13819
Epoch 320/350 Loss 0.13426
Epoch 330/350 Loss 0.13056
Epoch 340/350 Loss 0.12707
Epoch 350/350 Loss 0.12378


Explicacion del paso a paso

.eval: cambia al modo evaluacion

softmax(): devuelve las probabilidades entre 0 y 1

torch.argmax: la clase con la mayor probabilidad

accuracy_score(x,y): compara ambas listas y devuelve la probabilidad
la prediccion de la red estara en valores entre 0 y 1, se lo multiplica *100 y se tendra la probabilidad exacta

In [ ]:
from sklearn.metrics import accuracy_score

def evaluate(x):
    model.eval()
    y_pred = model(x)  # predecir
    # calcular las clases predichas
    y_probas = softmax(y_pred)  # probabilidades entre 0 y 1
    return torch.argmax(y_probas, axis=1)  # clase con mayor probabilidad

y_pred = evaluate(torch.from_numpy(X_test).float().cuda())
accuracy_score(y_test, y_pred.cpu().numpy())

0.997

en este caso el entrenamiento fu exitoso, los valores estan variando entre 0.9962 y 0.997 lo cual el margen de error es menor al 1%